hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -186,6 +186,34 @@ object LowerTableIR {
           val env: Env[IR] = Env("row" -> row, "global" -> loweredChild.globals)
           loweredChild.copy(body = StreamMap(loweredChild.body, row.name, Subst(newRow, BindingEnv(env, scan = Some(env)))))
 
+        case t@TableKeyBy(child, newKey, isSorted: Boolean) =>
+          val loweredChild = lower(child)
+          val nPreservedFields = loweredChild.partitioner.kType.fieldNames
+            .zip(newKey)
+            .takeWhile { case (l, r) => l == r }
+            .length
+
+          require(!isSorted || nPreservedFields > 0 || newKey.isEmpty)
+
+          if (nPreservedFields == newKey.length)
+            loweredChild
+          else if (isSorted) {
+            val newPartitioner = loweredChild.partitioner
+              .coarsen(nPreservedFields)
+              .extendKey(t.typ.keyType)
+            if (newPartitioner.isStrict)
+              loweredChild.copy(partitioner = newPartitioner)
+            else throw new LowererUnsupportedOperation("TableKeyBy with isSorted but a non-strict partitioner requires partitioner adjustments")
+          } else
+            ctx.backend.lowerDistributedSort(loweredChild, newKey.map(k => SortField(k, Ascending)))
+
+        case TableOrderBy(child, sortFields) =>
+          val loweredChild = lower(child)
+          if (TableOrderBy.isAlreadyOrdered(sortFields, loweredChild.partitioner.kType.fieldNames))
+            loweredChild.copy(partitioner = loweredChild.partitioner.coarsen(0))
```

Comment:
I don't think you need to coarsen here.

Reply:
I guess I need to clarify what the relationship of a partitioner is to the key of a lowered TIR. Does the partitioner key type have to be the same? Longer?

Reply:
I though the partitioner could be longer (you potentially let it stay longer in the `nPreservedFields == newKey.length` case above), but it seems not.  From RVD:

 > require(typ.kType.virtualType isIsomorphicTo partitioner.kType)

(It seems isIsomorphicTo is equals for base structs: it allows structs and tuples to compare equal.)

So why isn't this `coarsen(sortFields.length)`?

Reply:
TableOrderBy can sort in descending order. We'll either need to encode ascending/descending into the partitioner, or erase sortedness when we order by (descending).

Reply:
er, nevermind, I think you're right -- no need to coarsen.